homeassistant/components/powerwall/config_flow.py

```diff
@@ -56,41 +62,101 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
 
     def __init__(self):
         """Initialize the powerwall flow."""
-        self.ip_address = None
+        self.ip_address: str | None = None
+        self.title: str | None = None
+        self.reauth_entry: config_entries.ConfigEntry | None = None
 
     async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
         """Handle dhcp discovery."""
         self.ip_address = discovery_info.ip
-        self._async_abort_entries_match({CONF_IP_ADDRESS: self.ip_address})
-        self.context["title_placeholders"] = {CONF_IP_ADDRESS: self.ip_address}
-        return await self.async_step_user()
+        gateway_din = discovery_info.hostname.upper()
+        # The hostname is the gateway_din (unique_id)
+        await self.async_set_unique_id(gateway_din)
+        self._abort_if_unique_id_configured(updates={CONF_IP_ADDRESS: self.ip_address})
+        for entry in self._async_current_entries(include_ignore=False):
+            if entry.data[CONF_IP_ADDRESS] == discovery_info.ip:
+                if entry.unique_id is not None and is_ip_address(entry.unique_id):
+                    if self.hass.config_entries.async_update_entry(
+                        entry, unique_id=gateway_din
+                    ):
+                        self.hass.async_create_task(
+                            self.hass.config_entries.async_reload(entry.entry_id)
+                        )
+                return self.async_abort(reason="already_configured")
+        self.context["title_placeholders"] = {
+            "name": gateway_din,
+            "ip_address": self.ip_address,
+        }
+        errors, info = await self._async_try_connect(
+            {CONF_IP_ADDRESS: self.ip_address, CONF_PASSWORD: gateway_din[-5:]}
+        )
+        if errors:
+            return await self.async_step_user()
```

Comment:
Why do we move to the user step if there's a connection error during discovery? Why not just abort?

Reply:
The password being the last 5 of the gateway din is the default. It may have changed so they need to configure it manually at that point

Reply:
I think it makes sense to abort if its not a password error though.